### ISSUE TITLE
Wikimedia DAG times out due to very large batches

### ISSUE THREAD

**Issue by stacimc** [WordPress/openverse-catalog#725]
## Description

Some recent DagRuns of the `wikimedia_commons_workflow` have timed out before completion in the `pull_data` step. It looks like this is happening due to encountering extremely large batches.

These show up in the logs as very long sections where the reported `gaicontinue` token is identical, and the `gucontinue` token is almost identical except for a changing integer suffix. Here's a very small snippet from a recent run:

```
[2022-09-07, 15:52:30 UTC] {wikimedia_commons.py:106} INFO - New continue token: {'gucontinue': 'Arrows-orphan.svg|arwiki|3120418', 'gaicontinue': '20220802024848|Lohjanjärvi_in_September.jpg', 'continue': 'gaicontinue||imageinfo'}
[2022-09-07, 15:52:30 UTC] {wikimedia_commons.py:130} INFO - Got 250 pages
[2022-09-07, 15:52:30 UTC] {wikimedia_commons.py:130} INFO - Got 250 pages
[2022-09-07, 15:52:30 UTC] {wikimedia_commons.py:130} INFO - Got 250 pages
[2022-09-07, 15:52:31 UTC] {wikimedia_commons.py:106} INFO - New continue token: {'gucontinue': 'Arrows-orphan.svg|arwiki|3125414', 'gaicontinue': '20220802024848|Lohjanjärvi_in_September.jpg', 'continue': 'gaicontinue||imageinfo'}
[2022-09-07, 15:52:31 UTC] {wikimedia_commons.py:130} INFO - Got 250 pages
[2022-09-07, 15:52:31 UTC] {wikimedia_commons.py:130} INFO - Got 250 pages
[2022-09-07, 15:52:31 UTC] {wikimedia_commons.py:130} INFO - Got 250 pages
[2022-09-07, 15:52:32 UTC] {wikimedia_commons.py:106} INFO - New continue token: {'gucontinue': 'Arrows-orphan.svg|arwiki|3127601', 'gaicontinue': '20220802024848|Lohjanjärvi_in_September.jpg', 'continue': 'gaicontinue||imageinfo'}
```

In the DagRun that example is taken from, that batch took almost 11.5 hours to process (from the first log containing the `Arrows-orphan` continue token, to the final `Found batchcomplete` log), but only 208 records were ultimately processed from that batch.

## Reproduction

Go to the Wikimedia DAG in production and observe logs for any of the recent failed `pull_data` tasks. The example here comes from processing for August 02, 2022. You should also be able to manually run the DAG with that `date` to reproduce the issue.

## Possible resolution

Some initial thoughts for ways to resolve this:

1. **Increase the timeout for Wikimedia's `pull_data` task across the board.** I'm concerned about doing this because Wikimedia _generally_ should complete very quickly. This could cause problems in reingestion, which tries to run many Wikimedia dagruns.
2. **Add some handling for long-running batches.** We could try to add a custom timeout to Wikimedia's batch processing, to exit early if a batch like this is encountered. If we do so, we risk losing some data.

As part of this ticket it would be a good idea to look through the logs of a few more of these failures to try to confirm how much data we're getting from these large batches.

## Resolution

- [ ] 🙋 I would be interested in resolving this bug.

**Comment by stacimc** [comment]
For investigating this, it might also be a good idea to pull the code before the recent Wikimedia refactor and try running the DAG for a problematic date. I don't believe there were any changes to the continue token logic that should affect this, but it may be worth confirming.

**Comment by stacimc** [comment]
I've set the priority to `low`, as the Wikimedia DAG is still running. We are getting full data for most runs, and partial data for runs with the batch issue.

**Comment by AetherUnbound** [comment]
> **Increase the timeout for Wikimedia's `pull_data` task across the board.** I'm concerned about doing this because Wikimedia _generally_ should complete very quickly. This could cause problems in reingestion, which tries to run many Wikimedia dagruns.

Presently Wikimedia's pull timeout is 12 hours:

https://github.com/WordPress/openverse-catalog/blob/6e9d02d65ef42b92bcbc63d7cb1695d13318517d/openverse_catalog/dags/providers/provider_workflows.py#L198

I would also be concerned bumping this up too much further because it's an `@daily` DAG. If the runs end up taking 24 hours or more, we'll reach a saturation point with the Wikimedia DAG 😞

**Comment by AetherUnbound** [comment]
For testing purposes, runs on the following days yielded large-batch files:
* 2022-08-02
* 2022-08-08
* 2022-08-21
* 2022-09-09

**Comment by stacimc** [comment]
I've increased the priority to `medium` as this affects reingestion a bit more severely than it does ingestion (since reingestion timeouts are necessarily lower). I do not think this needs to be `high` as we are getting partial data.

**Comment by AetherUnbound** [comment]
Okay, I've uncovered why this is happening 😄 The short answer is that it's a result of us trying to get all of the global usages for media items that are super common (e.g. Flag of India) in an attempt to use that as a proxy for media popularity. The good news is that we _can_ move past those larger batches, the bad news is that it's quite complicated to do so. I'm working on a solution for this now, and I'll document everything I've uncovered (because Wikimedia is truly a beast!)
